tiri/src/main.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Tiri -- TIR interpreter.
+//!
+//! No effort has been made to make this fast.
+
+#[macro_use]
+extern crate log;
+
+use elf;
+use fallible_iterator::FallibleIterator;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
+use ykpack::{BasicBlockIndex, Body, Decoder, DefId, Pack, Statement, StatementIndex, Terminator};
+
+/// The interpreter maintains a stack of these to keep track of calls.
+struct Frame<'p> {
+    /// A reference to the TIR body corresponding with this `Frame`.
+    body: &'p Body,
+    /// The current execution position in the above TIR body.
+    pc: (BasicBlockIndex, StatementIndex),
+}
+
+impl<'p> Frame<'p> {
+    /// Create a new frame and position its program counter at the beginning of the first block.
+    fn new(body: &'p Body) -> Self {
+        Self { body, pc: (0, 0) }
+    }
+}
+
+/// The interpreter itself.
+/// The struct itself holds only immutable program information.
+struct Interp {
+    /// Maps a `DefId` to the corresponding TIR body.
+    tir_map: HashMap<DefId, Body>,
+    /// The `DefId` of the `main()` function, which serves as the entry point.
+    main_defid: DefId,
+}
+
+impl Interp {
+    /// Create a new interpreter, using the TIR found in the `.yk_tir` section of the binary `bin`.
+    fn new(bin: &Path) -> Self {
+        let ef = elf::File::open_path(&PathBuf::from(bin)).unwrap();
```

Comment:
Is `ef` needed as in intermediary? `let sec = elf::File::open_path(...).unwrap().get_section(".yk_tir").unwrap()` might be more succinct.

Reply:
Sadly, if we do:
```
let sec = elf::File::open_path(&PathBuf::from(bin)).unwrap().get_section(".yk_tir").unwrap();
```

We get:
```
error[E0716]: temporary value dropped while borrowed
  --> tiri/src/main.rs:52:19
   |
52 |         let sec = elf::File::open_path(&PathBuf::from(bin)).unwrap().get_section(".yk_tir").unwrap();
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                - temporary value is freed at the end of this statement
   |                   |
   |                   creates a temporary which is freed while still in use
53 |         let mut curs = Cursor::new(&sec.data);
   |                                    --------- borrow later used here
```

Reply:
Ah, OK.

Reply:
I was kind of hoping the non-linear lifetimes work would render this kind of error a thing of the past.

Perhaps that work never hit Rust?

Reply:
NLL is largely addressing a different issue, whereby the borrow checker rejects programs that appear to borrow a variable mutably and immutably. Here you've got a different problem, which is a temporary being freed while still being borrowed. I agree that this often feels quite annoying but I'm not sure if there's a sensible resolution or not.